File: utf64.test.ts
import { assert, test } from "vitest";
import { invalid_utf64, valid_utf64 } from "./test.json";
import { str_to_utf64, utf64_to_str } from "./utf64";

for (const [dest, src] of Object.entries(valid_utf64)) {
  test(dest, () => {
    assert.equal(dest, str_to_utf64(src));
    assert.equal(src, utf64_to_str(dest));
  });
}

for (const [dest, reason] of Object.entries(invalid_utf64)) {
  test(reason, () => {
    assert.throws(() => {
      const src = utf64_to_str(dest);
      console.log(dest, "->", src);
    });
  });
}

File: utf64_test.go
package utf64

import (
	"encoding/json"
	"os"
	"testing"
)

type TestJson struct {
	Valid   map[string]string `json:"valid_utf64"`
	Invalid map[string]string `json:"invalid_utf64"`
}

func TestUtf64(t *testing.T) {
	file, err := os.ReadFile("test.json")
	if err != nil {
		panic(err)
	}

	var spec TestJson
	err = json.Unmarshal(file, &spec)
	if err != nil {
		panic(err)
	}

	for dest, src := range spec.Valid {
		actualDest := Encode(src)
		if actualDest != dest {
			t.Errorf("\nWhen encoding: %v\nExpected: %v\nActual %v", src, dest, actualDest)
		}
		actualSrc, err := Decode(dest)
		if err != nil {
			t.Errorf("\nWhen decoding: %v\nUnexpected error: %v", dest, err)
		}
		if actualSrc != src {
			t.Errorf("\nWhen decoding: %v\nExpected: %v\nActual %v", dest, src, actualSrc)
		}
	}

	for dest, message := range spec.Invalid {
		actualSrc, err := Decode(dest)
		if err == nil {
			t.Errorf("\nWhen decoding: %v\nExpected error: %v\nBut got result: %v", dest, message, actualSrc)
		}
	}
}
